### ISSUE TITLE
v7 with experimental glx: window contents are not drawn beyond framebuffer size

### ISSUE THREAD

**Issue by Vladimir-csp** [yshui/picom#212]
### Platform

Debian testing.
### GPU, drivers, and screen setup
`OpenGL renderer string: Mesa DRI Intel(R) HD Graphics 520 (Skylake GT2)`
modesetting driver
### Environment
Openbox
### Compton version
`v6-rc1-226-g6cb5f97` (buit from stable/v7)

### Compton configuration:
```
shadow = true;
shadow-radius = 3;
shadow-offset-x = -3;
shadow-offset-y = -3;
shadow-opacity = 0.7;
shadow-red = 0.0;
shadow-green = 0.0;
shadow-blue = 0.0;
shadow-exclude = [
	"name = 'Notification'",
	"class_g = 'Conky'",
	"class_g ?= 'Notify-osd'",
	"class_g = 'Cairo-clock'",
	"_GTK_FRAME_EXTENTS@:c"
];
xinerama-shadow-crop = true;
shadow-ignore-shaped = true;

inactive-opacity = 1.0;
active-opacity = 1.0;
frame-opacity = 0.9;
inactive-opacity-override = false;
inactive-dim = 0.0;

blur:
{
	method = "box",
	size = 4;
#	deviation = 5.0;
}
blur-background-exclude = [
	"window_type = 'dock'",
	"window_type = 'desktop'",
	"window_type = 'dnd'",
	"_GTK_FRAME_EXTENTS@:c"
];

fading = true;
fade-in-step = 0.1;
fade-out-step = 0.1;
fade-exclude = [ ];
fade-delta = 15;

focus-exclude = [ "class_g = 'Cairo-clock'" ];
mark-wmwin-focused = true;
mark-ovredir-focused = true;
# use-ewmh-active-win = true;
detect-rounded-corners = true;
detect-client-opacity = true;

unredir-if-possible = true;
# unredir-if-possible-delay = 5000;
# unredir-if-possible-exclude = [ ];

backend = "glx";
vsync = true;
refresh-rate = 0;
detect-transient = true;
detect-client-leader = true;
invert-color-include = [ ];
# resize-damage = 1;

# GLX backend
glx-no-stencil = true;
glx-no-rebind-pixmap = true;
use-damage = true;
xrender-sync-fence = true;

wintypes :
{
	tooltip = { fade = true; shadow = true; opacity = 0.9; focus = true; full-shadow = false; };
	dock = { shadow = false; }
	dnd = { shadow = false; }
	popup_menu = { opacity = 0.9; }
	dropdown_menu = { opacity = 0.9; }
};

```

### Steps of reproduction

1. start `compton --experimental-backends`
2. stretch a window to be bigger than screen size.

### Expected behavior
Contents of a window that currently fit in the screen are shown.

### Current Behavior
Only upper-left screen-sized portion of window contents is rendered, regardless of window position.

**Comment by yshui** [comment]
Cannot reproduce this on NVIDIA 1050M. Might be a driver problem?

**Comment by Vladimir-csp** [comment]
Might be, but with standard glx backend this does not happen, only experimental.

**Comment by yshui** [comment]
experimental glx render things differently, probably that triggered the driver bug?

can you capture an apitrace of compton showing this problem?

thanks.

**Comment by Vladimir-csp** [comment]
Here is the trace:
[compton.trace.zip](https://github.com/yshui/compton/files/3463595/compton.trace.zip)
I launched featherpad, moved window beyond upper left corner of the display, then stretched it down to bottom right corner.

Also apitrace spammed this to console:
```
apitrace: warning: glXChooseFBConfig: unknown key 0x20B2, interpreting value as int
apitrace: warning: glXChooseFBConfig: unknown key 0x20B2, interpreting value as int
apitrace: warning: glXChooseFBConfig: unknown key 0x20B2, interpreting value as int
apitrace: warning: glXChooseFBConfig: unknown key 0x20B2, interpreting value as int
apitrace: warning: glXChooseFBConfig: unknown key 0x20B2, interpreting value as int
```

**Comment by mighty9245** [comment]
can reproduce with your config on archlinux running compton v7.4 in bspwm, can't reproduce with the default one.
`OpenGL renderer string: Mesa DRI Intel(R) UHD Graphics 620 (Kabylake GT2)`

**Comment by mighty9245** [comment]
investigated on this a little and looks like the `frame-opacity` option causes the issue, because commenting it out from the provided config makes the issue to disappear

**Comment by mighty9245** [comment]
hey, update on this: my issue with misplaced transparent lines on window instead of it's border is fixed with some commit. 

but the main issue is still here: `picom --experimental-backends --backend glx --config /dev/null --frame-opacity 0.5` and stretch window to be bigger than the screen size and you'll see that right part of the window isn't drawn.

this seems to be not a driver bug because i've just built picom and tested it on my main machine with an amd gpu (amdgpu driver) and issue is still here.

`OpenGL renderer string: AMD Radeon (TM) R9 380 Series (TONGA, DRM 3.36.0, 5.5.9-arch1-2, LLVM 9.0.1)`
```
nu:~$ pacman -Qs amdgpu
local/xf86-video-amdgpu 19.1.0-1 (xorg-drivers)
    X.org amdgpu video driver
```

you can see shadow on the part that is not drawn. btw idk if error in terminal is related, i just kill picom with ctrl-c and it pops up.
![image](https://user-images.githubusercontent.com/52252283/76967415-788fda80-6938-11ea-9328-6ea43160c245.png)

**Comment by mighty9245** [comment]
jesus christ i managed to fix this
![image](https://user-images.githubusercontent.com/52252283/77093798-959dd980-6a1c-11ea-9aff-2becae2f2164.png)

**Comment by yshui** [comment]
@mighty9245 what fixed it?

**Comment by mighty9245** [comment]
@mighty9245, some code modifications: basically you need to set viewports when dealing with framebuffers and also projection matrices for shaders if you're using them. 

i want to make my code look tidy and upload it in forked repository cuz i'm too shy to make a pull request since i'm not sure if my actions are right and there is no way to do the same in a better way.

**Comment by mighty9245** [comment]
@yshui, fix: f14bf3f542634ab2b5087ed55d706fb24b1f62ce

**Comment by yshui** [comment]
@Vladimir-csp this might have been fixed by #388. can you get the latest `next` and give it a go?

(Thank @mighty9245 and @tryone144)

**Comment by Vladimir-csp** [comment]
The problem is gone, thank you
